### ISSUE TITLE
[BR]: False positives when there is no end of line at the moment of reading.

### ISSUE THREAD

**Issue by markmal** [fail2ban/fail2ban#3223]
### Environment: 
- Fail2Ban version  : 0.11.2-1~upstream1
- OS, including release name/version : Debian 10 Linux. Linux  4.19.0-18-cloud-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux
- [ ] Fail2Ban installed via OS/distribution mechanisms: installed fail2ban_0.11.2-1.upstream1_all.deb
- [x] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

When at the moment of reading there is IP, timestamp, and partial message like this '185.202.220.105 - - [17/Feb/2022:04:13:23 -0500] "GET /res', or no message yet like this '185.202.220.105 - - [17/Feb/2022:04:13:23 -0500', fail2ban does not check that end of line is missing and processes the unfinished log record resulting a false match.

#### Steps to reproduce

Put unfinished record (without EOL) into end of test.log and test it with fail2ban-regex

#### Expected behavior

fail2ban should check that end of line is missing in a phase when it parses IP, timestamp, and the message after that, even it is end of file, and should keep reading until EOL appears. 

#### Observed behavior

False positives produced when a log record is not completed by a writer, only IP and timestamp, or partial message (and no EOL) is there at the moment of reading by fail2ban.

##### fail2ban.conf
```
loglevel = HEAVYDEBUG
```
##### Jail
```
[tomcat-app]
enabled  = true
port     = http,https,8080,8443
filter   = tomcat-app
logpath  = /var/log/tomcat/app_access.log
backend  = pyinotify
journalmatch =
maxretry = 5
```
##### Filter
```
[INCLUDES]

[Definition]

failregex = ^<HOST> -.*

ignoreregex = ^<HOST> -.* "(GET|POST|PUT|DELETE|HEAD)\s\/(|assets|csrf|quote|rest|user)[ \/?;].*
		^<HOST> -.* "-" 400 - "-" "-"
```

#### Relevant parts of /var/log/fail2ban.log file:
```
2022-02-17 04:07:27,586 fail2ban.filter         [16070]: TRACE   Working on line '185.202.220.105 - - [17/Feb /2022:04:07:26 -0500'
2022-02-17 04:07:27,586 fail2ban.datedetector   [16070]: HEAVY   try to match time for line: 185.202.220.105 
- - [17/Feb/2022:04:07:26 -0500
2022-02-17 04:07:27,587 fail2ban.filter         [16070]: HEAVY   Looking for match of [('185.202.220.105 - - 
[', '17/Feb/2022:04:07:26 -0500', '')]
2022-02-17 04:07:27,587 fail2ban.filter         [16070]: TRACE     Matched failregex 0: {'ip4': '185.202.220.
105', 'ip6': None, 'dns': None}
2022-02-17 04:07:27,587 fail2ban.filter         [16070]: DEBUG   Processing line with time:1645088846 and ip:
185.202.220.105
2022-02-17 04:07:27,587 fail2ban.filter         [16070]: INFO    [tomcat-bunzl-vip] Found 185.202.220.105 - 2022-02-17 04:07:26
```

#### Relevant lines from monitored log files:
```
185.202.220.105 - - [17/Feb/2022:04:07:26 -0500] "GET /rest/item?companyNumber=43&location=430&partNumber=L72700015 HTTP/1.1" 200 2172 "-" "python-requests/2.27.1"
```

**Comment by sebres** [comment]
> fail2ban does not check that end of line(EOL) is missing

1. see [#2766](https://github.com/fail2ban/fail2ban/issues/2766#issuecomment-650199529) for the answer why such logging is worse;
2. this behavior is changed in v.1.0 starting with 996920cdaa7b678afc64d1bef6a07ed857c81468 and later, it will now wait for line end:
https://github.com/fail2ban/fail2ban/blob/246d0e1100943e35c9e799463f152621bd35a2b2/fail2ban/server/filter.py#L1319

> and processes the unfinished log record resulting a false match.

No. It is not just unfinished log that resulting to false match, but your regular expression, which is vulnerable because of quasi unanchored matter and catch-alls inside. Rewrite your failregex more precise, fully anchored and without catch-alls and it'd stop...Or upgrade to v.1.0, but better don't log unbuffered (see #2766 again).
Another variant would be to switch to systemd backend (to monitor systemd journals if your application supporting that).

**Comment by markmal** [comment]
I've created FIFO pipe and configured Tomcat to write log into it.
Another process reads from the pipe and sends that to systemd journal.
```
(cat app_access.fifo | systemd-cat -t app_access) &
```
The problem is with dates now. Systemd has its own timestamp, and Tomcat writes its one, that is more precise, and is date of the actual HTTP request.
```
Feb 19 14:29:56 my.server.com app_access[7486]: 2022-02-19 14:29:48.731-05:00 104.158.244.72 - - "GET /wrong3 HTTP/2.0"
```
I think it would be better if F2B uses Tomcat's timestamp. How to specify which timestamp F2B should use. 
There is <HOST> var, is there a <TIMESTAMP> var that I can put into proper place in regex?

**Comment by sebres** [comment]
> But we cannot do anything with it because it is Tomcat logging.

I don't think so. Moreover I'm pretty sure that although apache FileHandler's supports buffering of the logs, the buffering is not enabled by default. So you  should check what is specified in conf and remove or comment out lines like `org.apache.juli.FileHandler.bufferSize`.

> May be not, for performance ...

Performance in logger worker? It is fast enough to write line-buffered log faster than whole pool of workers may ever process the requests.
Anyway buffering of logging info is simply a reason of latency as well as of issues like this.

> so atomicity of a "write record" operation is not guaranteed by OS/file system.

This is not a matter of atomic write, rather a matter of line-buffered handler (which would flush every single line) and atomic is mostly guaranteed by a queue for log messages of single writer or the mutex by multiple handlers.

> By the way I've looked into the code and saw this before I've opened the case. 

This is in source code of master branch (v.1.0) solely.

> I've decided the functionality is there but it does not work for some reason.

Sure it doesn't work, it is not in 0.11.x (only in 1.0) and you seem to use 0.11.2.

> I think it would be better if F2B uses Tomcat's timestamp.

Well, basically not - first datepattern is anchored at begin of message so it is significantly faster.
But you don't need to match it at all with systemd backend, because fail2ban already has the timestamp from journal, so the message line supplied to filter starts with `my.server.com app_access[7486]: ...` (there is no first time at all).
And the second time can be simply ignored with this RE (optional so can be safely ignored if your web-app could write log without extra timestamp) - `(?:2\d{3}-\d{2}-\d{2} \d{2}:\S+ )?`
The only small deviation is that it is the time of response (end of request) + some small delay by redirect to journal, compared to start time of request in case of your web-app. But if this deviation is acceptable, it'd not bother fail2ban at all.

So your filter (or jail) may look like this:
```ini
failregex = ^\s*\S+ app_access\[\d+\]: (?:2\d{3}-\d{2}-\d{2} \d{2}:\S+ )?<ADDR> [^"]*"[A-Z]+ /(?!assets|csrf|quote|rest|user)\S+ [^"]*"
#journalmatch = _SYSTEMD_UNIT=app_access.service + _COMM=app_access
journalmatch = SYSLOG_IDENTIFIER=app_access
# for file related backends only
datepattern = {^LN-BEG}
```
where `(?!...)` in failregex is a negative lookahead which would avoid URIs from your first line of `ignoreregex` from issue description (and empty URI is ignored by `\S+` after `/` with negative lookahead).
I don't understand your second `ignoreregex` at all (`^<HOST> -.* "-" 400 - "-" "-"`), so I don't know what exactly it is intended to ignore. 
If 400 is a HTTP response code (but why one would ignore "bad request"), then your second `ignoreregex` is not (always) correct... And not really expected since URI-part `-` will be anyway not matched by precise RE matching only `/...`.
But you could write it also as single RE without ignoreregex:
```ini
failregex = ^\s*\S+ app_access\[\d+\]: (?:2\d{3}-\d{2}-\d{2} \d{2}:\S+ )?<ADDR> [^"]*"[A-Z]+ /(?!assets|csrf|quote|rest|user)\S+ [^"]*"\s(?!400\s)\d+\s
```
where RE `\s(?!400\s)\d+\s` would match any response code (after space) excepting 400.
